plookup/src/table/hash_tables/tables.rs

```diff
@@ -0,0 +1,207 @@
+/// The result is another vector y in (F_p)^t.
+#[derive(Debug)]
+pub struct HashTable {
+    pub first_rows: Vec<[BlsScalar; 4]>,
+    pub middle_rows: Vec<[BlsScalar; 4]>,
+    pub end_rows: Vec<[BlsScalar; 4]>,
+}
+
+impl HashTable {
+    /// Create a new hash table and makes
+    /// empty vectors for each field.
+    pub fn new() -> Self {
+        Self {
+            first_rows: vec![],
+            middle_rows: vec![],
+            end_rows: vec![],
+        }
+    }
+
+    // The whole lookup table will be constructed in 3 parts: the first rows where the third
+    // entry is derived from the function F, i.e. the rows are of the form (_, _, F(i), ...).
+    // The middle rows are where the first entries are between V+1 and s_i for some i.
+    // The binary rows are at the bottom of the table, and they enumerate all binary possibilities
+    // on T_S bits.
+    fn f_rows(&mut self, f: Polynomial) {
+        for i in 0..(V + 1) {
+            let eval = f.evaluate(&BlsScalar::from(i as u64));
+            let row = [
+                BlsScalar::from(i as u64),
+                BlsScalar::zero(),
+                eval,
+                -BlsScalar::one(),
+            ];
+
+            self.first_rows.push(row);
+        }
+    }
+
+    // This function fills in the middle section of the hash table
+    // where the entry is defined as being between V+1 and s_i
+    // for a chosen i. The i here depends on the intialisation
+    // of the first rows.
+    fn m_rows(&mut self) {
+        for i in 0..N {
+            let distance = S[i as usize] - V as u64;
+
+            for j in 1..(distance + 1) {
+                let row = [
+                    BlsScalar::from(V as u64 + j),
+                    BlsScalar::from(i as u64 + 1),
+                    BlsScalar::from(V as u64 + j),
+                    -BlsScalar::one(),
+                ];
+
+                self.middle_rows.push(row);
+            }
+        }
+    }
+
+    // A function that creates all binary values of word of length T_S.
+    // For a width T_S, these are the binary possibilites for numbers
+    // between 0 and 15.
+    fn binary_end_rows(&mut self) {
+        let mut row = [BlsScalar::zero(); 4];
+        self.end_rows.push(row);
+
+        for _ in 1..(2i32.pow(T_S.try_into().unwrap())) {
+            incrementer(&mut row, 0);
+            self.end_rows.push(row);
+        }
+    }
+
+    /// This function constructs a hash table based on the
+    /// constants declared.
+    pub fn construct_table(f: Polynomial) -> Self {
+        let mut table = HashTable::new();
+        table.f_rows(f);
+        table.m_rows();
+        table.binary_end_rows();
+
+        table
+    }
+}
+
+/// The binary end rows section of the a hash table requires
+/// a function which fills out the whole vector, of arity 4,
+/// dependant on the given initial entry.
+pub fn incrementer(mut row: &mut [BlsScalar; 4], i: usize) {
```

Comment:
if `i > 4` this  will panic. Impl error handling or mention in the docs that this can panic.

Reply:
Still not addressed